asmcomp/asmlink.ml

```diff
@@ -178,30 +178,42 @@ let read_file obj_name =
   end
   else raise(Error(Not_an_object_file file_name))
 
-let scan_file obj_name tolink = match read_file obj_name with
+let scan_file obj_name (tolink, objfiles_rev) = match read_file obj_name with
   | Unit (file_name,info,crc) ->
       (* This is a .cmx file. It must be linked in any case. *)
       remove_required info.ui_name;
       List.iter (add_required file_name) info.ui_imports_cmx;
-      (info, file_name, crc) :: tolink
+      (info, file_name, crc) :: tolink, obj_name::objfiles_rev
   | Library (file_name,infos) ->
       (* This is an archive file. Each unit contained in it will be linked
          in only if needed. *)
       add_ccobjs (Filename.dirname file_name) infos;
-      List.fold_right
-        (fun (info, crc) reqd ->
-           if info.ui_force_link
-             || !Clflags.link_everything
-             || is_required info.ui_name
-           then begin
-             remove_required info.ui_name;
-             List.iter (add_required (Printf.sprintf "%s(%s)"
-                                        file_name info.ui_name))
-               info.ui_imports_cmx;
-             (info, file_name, crc) :: reqd
-           end else
-             reqd)
-        infos.lib_units tolink
+      let tolink =
+        List.fold_right
+          (fun (info, crc) reqd ->
+             if info.ui_force_link
+               || !Clflags.link_everything
+               || is_required info.ui_name
+             then begin
+               remove_required info.ui_name;
+               List.iter (add_required (Printf.sprintf "%s(%s)"
+                                          file_name info.ui_name))
+                 info.ui_imports_cmx;
+               (info, file_name, crc) :: reqd
+             end else
+               reqd)
+          infos.lib_units tolink
+      in
+        if Config.ccomp_type = "msvc"
+        && infos.lib_units = []
+        && not (Sys.file_exists (object_file_name obj_name)) then
```

Comment:
It would make sense to just test `infos.lib_units = []`.  Even with non-MSVC toolchains, it makes sense to not pass an empty library to the linker, resulting in a shorter command line and less work for the linker.

As to testing that the file doesn't exist, this is to be commended as very prudent, but 1- it causes one extra system call per file to link, and 2- I can't imagine a scenario where object files would be added by hand to the library file that goes with an OCaml `.cmxa` file.

Reply:
I have no problem with doing that - the test I've put is just belt-and-braces to fix exactly the bug. The the `.a` file may have been subsequently altered (e.g. to add stub objects directly) - I've no knowledge of somewhere doing this, but do we definitely want to break some undefined behaviour while fixing this particular bug?

I did also consider whether it would be better to add a field to the .cmxa format which explicitly specifies the object to link - so on MSVC for an empty library, this field would explicitly be blank. In principle, .cmx format could do the same thing, although a `.o` and `.obj` file is always associated, even for an empty `.ml` file.

Reply:
I (potentially temporarily) left this as-is. This change alters this PR from a bugfix to a potentially breaking change (albeit by using largely unspecified behaviour of the linker).

Reply:
Am I reading right that the third part of the test is here to protect users that were creating a `.lib` file by hand  behind the back on the compiler on msvc? Do we really intend to support this use case?

Reply:
@Octachron - it's more that I don't wish to intentionally break it in this PR. I believe there are some Mirage libraries which do this (although none which would hit this explicit case - i.e. an empty library with objects being added)